src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -281,6 +284,96 @@ private static void _ValidateAdditionalNIntDefinitions()
                         pInputBuffer += 4;
                         tempUtf16CodeUnitCountAdjustment -= 2; // 4 UTF-8 code units -> 2 UTF-16 code units (and 2 scalars)
 
+                        if (Sse2.IsSupported && Bmi1.IsSupported)
+                        {
+                            // Optimization: If we see 2-byte subsequences in the text, perhaps the text consists
+                            // mostly of 2-byte subsequences with some scattered ASCII.
+
+                            Vector128<sbyte> vecAll61 = Vector128.Create(unchecked((sbyte)0x61));
+                            Vector128<sbyte> vecAllA0 = Vector128.Create(unchecked((sbyte)0xA0));
+                            Vector128<sbyte> vecAllBF = Vector128.Create(unchecked((sbyte)0xBF));
+
+                            // The check below is really "can we read 16 bytes from the input buffer?" since 'pFinalPos...' is already offset
+                            // n.b. The subtraction below could result in a negative value (since we advanced pInputBuffer above), so
+                            // use nint instead of nuint.
+
+                            while ((nint)(pFinalPosWhereCanReadDWordFromInputBuffer - pInputBuffer) >= 12)
+                            {
+                                // Note the below is an UNALIGNED load. We'll also pre-populate 'thisDWord' just
+                                // in case we need to jump out of the loop eagerly.
+
+                                Vector128<sbyte> thisXmm = Sse2.LoadVector128((sbyte*)pInputBuffer);
+                                thisDWord = Sse2.ConvertToUInt32(thisXmm.AsUInt32());
+
+                                // The vector below is a mask of two-byte header bytes [ C2..DF ]
+                                // which are followed by a continuation byte [ 80..BF ].
+
+                                Vector128<sbyte> validTwoByteStarts =
+                                    Sse2.AndNot(
+                                        Sse2.CompareGreaterThan(Sse2.ShiftRightLogical128BitLane(thisXmm, 1), vecAllBF),
+                                        Sse2.CompareGreaterThan(Sse2.Add(thisXmm, vecAllA0), vecAll61));
```

Comment:
This logic could do with a bit more commenting. It is potentially overflowing each byte and comparing against 0x61, which basically just looks like magic, but as I understand it, it is effectively just trying to do an unsigned range check and accounting for the fact the comparison is signed😄